internal/dataplane/dataplane_tpl.go

```diff
@@ -0,0 +1,226 @@
+	credentials     *credentialsConfig
+}
+
+type consulConfigTmplArgs struct {
+	Addresses       string
+	GRPCPort        int
+	SkipServerWatch bool
+	TLSJSON         string
+	CredentialsJSON string
+}
+
+func (c consulConfig) generateJSON() (string, error) {
+	tmplArgs := &consulConfigTmplArgs{
+		Addresses:       c.addresses,
+		GRPCPort:        c.grpcPort,
+		SkipServerWatch: c.skipServerWatch,
+	}
+
+	if c.tls != nil {
+		tmplArgs.TLSJSON = c.tls.generateJSON()
+	}
+
+	if c.credentials != nil {
+		tmplArgs.CredentialsJSON = c.credentials.generateJSON()
+	}
+
+	t, err := template.New("consulConfig").Parse(consulConfigTemplate)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	err = t.Execute(&buf, tmplArgs)
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), err
+}
+
+const consulConfigTemplate = `{
+	"addresses": "{{ .Addresses }}",
+	"grpcPort": {{ .GRPCPort }},
+	"serverWatchDisabled": {{ .SkipServerWatch }}
+	{{- if .TLSJSON }}
+	,
+	"tls": {{ .TLSJSON }}
+	{{- end }}
+	{{- if .CredentialsJSON }}
+	,
+	"credentials": {{ .CredentialsJSON }}
+	{{- end }}
+}`
+
+type tlsConfig struct {
+	grpcCACertPath string
+	tlsServerName  string
+}
+
+func (t *tlsConfig) generateJSON() string {
+	return `{
+		"disabled": false,
+		"caCertsPath": "` + t.grpcCACertPath + `",
+		"tlsServerName": "` + t.tlsServerName + `"
+	}`
+}
+
+type credentialsConfig struct {
+	credentialType string
+	static         staticCredentialConfig
+}
+
+func (c *credentialsConfig) generateJSON() string {
+	return `{
+		"type": "` + c.credentialType + `",
+		"static":` + c.static.generateJSON() + `
+	}`
+}
+
+type staticCredentialConfig struct {
+	token string
+}
+
+func (s staticCredentialConfig) generateJSON() string {
+	return `{
+		"token": "` + s.token + `"
+	}`
+}
+
+type serviceConfig struct {
+	nodeName       string
+	proxyServiceID string
+	namespace      string
+	partition      string
+}
+
+func (s serviceConfig) generateJSON() string {
+	return `{
+		"nodeName": "` + s.nodeName + `",
+		"serviceID": "` + s.proxyServiceID + `",
+		"namespace": "` + s.namespace + `",
+		"partition": "` + s.partition + `"
+	}`
+}
+
+type loggingConfig struct {
+	json  bool
+	level string
+}
+
+func (l loggingConfig) generateJSON() string {
+	return `{
+		"logLevel": "` + l.level + `",
+		"logJSON": ` + strconv.FormatBool(l.json) + `
+	}`
+}
+
+type xdsServerConfig struct {
+	address string
+	port    int
+}
+
+func (x xdsServerConfig) generateJSON() string {
+	return `{
+		"bindAddress": "` + x.address + `",
+		"bindPort": ` + strconv.FormatInt(int64(x.port), 10) + `
+	}`
+}
+
+type envoyConfig struct {
+	adminBindAddress string
+	adminBindPort    int
+}
+
+func (e envoyConfig) generateJSON() string {
+	return `{
+		"adminBindAddress": "` + e.adminBindAddress + `",
+		"adminBindPort": ` + strconv.FormatInt(int64(e.adminBindPort), 10) + `
+	}`
+}
+
+type telemetryConfig struct {
+	useCentralConfig bool
+}
+
+func (t telemetryConfig) generateJSON() string {
```

Comment:
Should we add prometheus related stuff to this JSON? I see that those settings are overridable by ENV vars and so decided to leave it like this with the value of `useCentralConfig` as false

Reply:
I actually think we should not set default values here that are different than consul-dataplane 🤔 (except where it really makes sense to do otherwise).

So I'd opt to remove the telemetryConfig here, and consul-dataplane will default to `true` for this.

(I know I wrote in the RFC that this would default to false, but that was before consul-dataplane was actually implemented, and we ended up with `-use-central-config=true` as the default in Consul Dataplane.)